### PR TITLE
[Feature] Support to query relation metrics through PromQL.

### LINKED ISSUE #12245: [Feature] Support to query relation metrics through PromQL
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

[PromQL Service](https://skywalking.apache.org/docs/main/next/en/api/promql-service/#supported-metrics-scopehttpsgithubcomapacheskywalkingtree19a2395a9ebbf7dca1bcf7aefba0c1819aad7214oap-serverserver-coresrcmainjavaorgapacheskywalkingoapservercorequeryenumerationscopejavacatalog) doesn't support metrics of relation between services, instances or endpoints.

### Use case

User could use this to build dashboards to show metrics from the lines on the topology or manually click.

### Related issues

_No response_

### Are you willing to submit a pull request to implement this on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yswdqz**
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #12245.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

**Comment by yswdqz** [comment]
bd5ade0bf1fc72f9f252931b67767c13
I have wrote a unit test, and I tested service_relation part on local grafana dashboard. It seems work.

File: docs/en/api/promql-service.md [hunk 1]
```diff
@@ -389,20 +389,23 @@ Not all scopes are supported for now, please check the following table:
 | Service                 | yes     |
 | ServiceInstance         | yes     |
 | Endpoint                | yes     |
-| ServiceRelation         | no      |
-| ServiceInstanceRelation | no      |
+| ServiceRelation         | yes     |
+| ServiceInstanceRelation | yes     |
```
Comment (wu-sheng) [inline docs/en/api/promql-service.md:393]: We have EndpointRelation too, I think the doc missed that before.

**Review by wankai123 — COMMENTED** [review]
please also add test cases in `promql` e2e.

File: oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java [hunk 2]
```diff
@@ -440,6 +440,32 @@ private Entity buildEntity(Layer layer,
                 checkLabels(labelMap, LabelName.ENDPOINT);
                 entity.setEndpointName(labelMap.get(LabelName.ENDPOINT));
                 break;
+            case ServiceRelation:
+                checkLabels(labelMap, LabelName.DEST_SERVICE);
+                checkLabels(labelMap, LabelName.DEST_LAYER);
```
Comment (wankai123) [inline oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java:445]:
```suggestion
                checkLabels(labelMap, LabelName.DEST_SERVICE, LabelName.DEST_LAYER);
```
We can check the labels at the same time.

File: oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java [hunk 3]
```diff
@@ -440,6 +440,32 @@ private Entity buildEntity(Layer layer,
...
+                entity.setDestServiceName(labelMap.get(LabelName.DEST_SERVICE));
+                entity.setDestNormal(Layer.nameOf(labelMap.get(LabelName.DEST_LAYER)).isNormal());
+                break;
+            case ServiceInstanceRelation:
+                checkLabels(labelMap, LabelName.DEST_SERVICE);
+                checkLabels(labelMap, LabelName.DEST_SERVICE_INSTANCE);
+                checkLabels(labelMap, LabelName.SERVICE_INSTANCE);
+                checkLabels(labelMap, LabelName.DEST_LAYER);
```
Comment (wankai123) [inline oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java:453]:
```suggestion
                checkLabels(labelMap, LabelName.DEST_SERVICE, LabelName.DEST_SERVICE_INSTANCE,  LabelName.SERVICE_INSTANCE, LabelName.DEST_LAYER);
```

File: oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java [hunk 4]
```diff
@@ -440,6 +440,32 @@ private Entity buildEntity(Layer layer,
...
+                entity.setServiceInstanceName(labelMap.get(LabelName.SERVICE_INSTANCE));
+                entity.setDestServiceInstanceName(labelMap.get(LabelName.DEST_SERVICE_INSTANCE));
+                entity.setDestServiceName(labelMap.get(LabelName.DEST_SERVICE));
+                entity.setDestNormal(Layer.nameOf(labelMap.get(LabelName.DEST_LAYER)).isNormal());
+                break;
+            case EndpointRelation:
+                checkLabels(labelMap, LabelName.DEST_SERVICE);
+                checkLabels(labelMap, LabelName.DEST_ENDPOINT);
+                checkLabels(labelMap, LabelName.ENDPOINT);
+                checkLabels(labelMap, LabelName.DEST_LAYER);
```
Comment (wankai123) [inline oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java:463]:
```suggestion
                checkLabels(labelMap, LabelName.DEST_SERVICE, LabelName.DEST_ENDPOINT, LabelName.ENDPOINT, LabelName.DEST_LAYER);
```

**Comment by wu-sheng** [comment]
You don't add source layer, I think?
Please order them in a certain order to avoid confusion.

Service + instance/endpoint + layer at both sides.

**Comment by yswdqz** [comment]
> You don't add source layer, I think? 

Source Layer has added at another code.

> Please order them in a certain order to avoid confusion.

Is the checklabel part?

**Comment by wu-sheng** [comment]
> > You don't add source layer, I think?
> 
> Source Layer has added at another code.

What do you mean other codes? I mean your new added codes. The previous ones are for other cases, aren't they?

> 
> > Please order them in a certain order to avoid confusion.
> 
> Is the checklabel part?

Yes. It helps for code reading.

**Comment by yswdqz** [comment]
> What do you mean other codes? I mean your new added codes. The previous ones are for other cases, aren't they?

Whatever the type,Layer has added at here.
https://github.com/apache/skywalking/blob/caeaec6387cf43bb8659238c18cc0c56b9226cc8/oap-server/server-query-plugin/promql-plugin/src/main/java/org/apache/skywalking/oap/query/promql/rt/PromQLExprQueryVisitor.java#L217

**Comment by wu-sheng** [comment]
I think once you fix another suggestion, and add dashboards for these relations, we're good.

**Comment by wu-sheng** [comment]
@yswdqz How about this?

**Comment by yswdqz** [comment]
> @yswdqz How about this?

I've been busy with a few things the last few days, so I'll try to get it done today

**Comment by yswdqz** [comment]
I found a bug about armeria,  it will delete `'`, and make promql antlr4 parser sytax mistake in this case.
image

**Comment by wu-sheng** [comment]
Why don't you use double quotations mark to separate? Is your way grammarly correct for a HTTP? 

I prefer the way is "xxxx='yyy'" or 'xxx="yyy"'

**Comment by yswdqz** [comment]
Is these OK? And I think this will be commit in showcase?
image
image
image
image
image
image
image

**Comment by wu-sheng** [comment]
As folded dashboards are added, you could only keep services and layers out of all dashboards, and move instances and endpoints into relative dashboard directly.

**Comment by wu-sheng** [comment]
And for metric graphs in the dashboard, you could make 3 widgets per row. One per row seems too wide, right?

**Comment by yswdqz** [comment]
> And for metric graphs in the dashboard, you could make 3 widgets per row. One per row seems too wide, right?

Endpoint only has 4 widgets, and service and serviceinstance has 8(server side and client side, so 4 rows is good), so I keep the same style with service and instance(4 rows).

**Comment by yswdqz** [comment]
> As folded dashboards are added, you could only keep services and layers out of all dashboards, and move instances and endpoints into relative dashboard directly.

Sorry, I am not familiar with grafana, how can I make it?

**Comment by wu-sheng** [comment]
Can't you move this into the instance tab?

image

**Comment by wu-sheng** [comment]
> > And for metric graphs in the dashboard, you could make 3 widgets per row. One per row seems too wide, right?
> 
> Endpoint only has 4 widgets, and service and serviceinstance has 8(server side and client side, so 4 rows is good), so I keep the same style with service and instance(4 rows).

Why you need to keep number of rows same? This page is long, we should keep the page as short as possible. Instance could try with 4 widget per row(2 for client, 2 for server) and same for the endpoint.

**Comment by yswdqz** [comment]
> Can't you move this into the instance tab?

I don't know how to do that. Services, ServiceInstances, Endpoints are variables in dashboard. And I don't know how can I bind it to pannel level (maybe not support?).

**Comment by yswdqz** [comment]
> Why you need to keep number of rows same? This page is long, we should keep the page as short as possible. Instance could try with 4 widget per row(2 for client, 2 for server) and same for the endpoint.

Is that OK?
image
image
image

**Comment by wu-sheng** [comment]
> > Can't you move this into the instance tab?
> 
> 
> 
> I don't know how to do that. Services, ServiceInstances, Endpoints are variables in dashboard. And I don't know how can I bind it to pannel level (maybe not support?).
> 
> 

If that is noy supported, I am good of the latest status.

**Review by wankai123 — APPROVED** [review]
LGTM

**Comment by wu-sheng** [comment]
We should have docs update to point to https://github.com/apache/skywalking-showcase/pull/176?

**Comment by yswdqz** [comment]
I didn't find the documentation I needed to change, the only documentation I found related to grafana was https://skywalking.apache.org/docs/main/next/en/setup/backend/ui-grafana/

**Comment by wu-sheng** [comment]
@wankai123 Any thought?

**Comment by wankai123** [comment]
https://skywalking.apache.org/docs/main/next/en/setup/backend/ui-grafana/  pointed to the Grafana Demo in the end section. Users can export the template from the demo or download it from the showcase.

But I think we could update/add a relation-metrics section under https://skywalking.apache.org/docs/main/next/en/setup/backend/ui-grafana/#configure-metric-dashboards

**Comment by wu-sheng** [comment]
@yswdqz When your showcase PR gets merged, could you update a list here pointing to all dashboards?

**Comment by yswdqz** [comment]
Got it.

**Pull request merged**